clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -522,7 +525,35 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener callb
 
     @Override
     public void assign(Collection<TopicPartition> partitions) {
-        throw new KafkaException("method not implemented");
+        if (partitions == null) {
+            throw new IllegalArgumentException("Topic partitions collection to assign to cannot be null");
+        }
+
+        if (partitions.isEmpty()) {
+            this.unsubscribe();
+            return;
+        }
+
+        for (TopicPartition tp : partitions) {
+            String topic = (tp != null) ? tp.topic() : null;
+            if (Utils.isBlank(topic))
+                throw new IllegalArgumentException("Topic partitions to assign to cannot have null or empty topic");
+        }
+        // TODO: implement fetcher
+        // fetcher.clearBufferedDataForUnassignedPartitions(partitions);
+
+        // make sure the offsets of topic partitions the consumer is unsubscribing from
+        // are committed since there will be no following rebalance
+        commit(subscriptions.allConsumed());
```

Comment:
I'll take a closer look at this.

Reply:
Hey @junrao - I believe we still want to commit the offset before there is a partition change. I think if we don't commit offset here, it is possible to lose some progress.

Reply:
@philipnee: `KafkaConsumer.assign()` has the following code. 

```
                // make sure the offsets of topic partitions the consumer is unsubscribing from
                // are committed since there will be no following rebalance
                if (coordinator != null)
                    this.coordinator.maybeAutoCommitOffsetsAsync(time.milliseconds());
```

1. In the old code, if the groupId is not specified, coordinator will be null and the offset commit will be skipped. In this PR, we always call offset commit.
2. In the old code, the implementation of `maybeAutoCommitOffsetsAsync()` is guarded by the `ENABLE_AUTO_COMMIT_CONFIG` config. In this PR, there is no such guard.

Are both changes expected?

Reply:
Hey @junrao - Sorry I misunderstood your concern.  To your question: If the coordinator is not available, the commit request manager won't be built and this commit() will be skipped.  In the `RequestManagers.java` we've got
```
if (groupRebalanceConfig != null && groupRebalanceConfig.groupId != null) {
                    final GroupState groupState = new GroupState(groupRebalanceConfig);
                    coordinator = new CoordinatorRequestManager(time,
                            logContext,
                            retryBackoffMs,
                            errorEventHandler,
                            groupState.groupId);
                    commit = new CommitRequestManager(time, logContext, subscriptions, config, coordinator, groupState);
                }
```

to your second question: autoCommit is still guarded by the config per this line in `CommitRequestManager.java`
```
private void maybeAutoCommit() {
   if (!autoCommitState.isPresent()) {
            return;
        }
   // autocommit otherwise
}
```

On the Consumer API level, we treat sync/async commit to be the same, except sync commit waits for the completion of the future.

Reply:
Thanks for the explanation, @philipnee.

I understand the first point now. 

For the 2nd point, `CommitRequestManager.maybeAutoCommit()` is guarded by the auto commit flag. However, this call is made only internally in `CommitRequestManager`. When processing a commit event from the application, it doesn't seem that the auto commit flag is checked.

Reply:
Hey @junrao - Thanks for catching this.  I made a minor change to the implementation that the application thread would send an event (with the topic partitions to commit) to the background thread to try to perform an autocommit.  I think now it has the same logic as the KafkaConsumer that it would check - 1. if the commit manager is there (per the explanation above), 2. autocommit is enabled or not, and 3. update the auto-commit timer and check if the timer is expired. If the timer has expired, send an auto-commit and reset the timer.

Reply:
I would suggest trying to get some tests similar to the existing [KafkaConsumerTest.testManualAssignmentChangeWithAutoCommitEnabled](https://github.com/apache/kafka/blob/cbd5d52ec7b89a5bbbbc245f5373a98d8379ab5a/clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java#L1411C17-L1411C64)
and [KafkaConsumerTest.testManualAssignmentChangeWithAutoCommitDisabled](https://github.com/apache/kafka/blob/cbd5d52ec7b89a5bbbbc245f5373a98d8379ab5a/clients/src/test/java/org/apache/kafka/clients/consumer/KafkaConsumerTest.java#L1465) to ensure the behaviour from the API level.

Reply:
Thanks, @lianetm - As autocommit is handled by the commitRequestManager, the application thread is only responsible for sending an assignment change event to the background thread.  This is tested in the `PrototypeAsyncConsumer.testAssign()`.  The autocommit was also tested in the request manager.  I would add a test to the DefaultBackgroundThreadTest to test if the AssignmentChange event triggers the autocommit.